crates/bevy_ecs/src/query/iter.rs

```diff
@@ -397,8 +316,11 @@ where
         QueryIterationCursor {
             fetch,
             filter,
-            table_id_iter: query_state.matched_table_ids.iter(),
-            archetype_id_iter: query_state.matched_archetype_ids.iter(),
+            id_iter: if Self::IS_DENSE {
```

Comment:
If we want to keep this, I want solid safety comments explaining what's happening here and why it's safe.

Reply:
Done.